## **Importazione librerie**


In questa fase vengono importate le librerie necessarie per la gestione, l'esplorazione e la pulizia del dataset. In particolare, pandas viene utilizzata per la manipolazione dei dati tabellari, mentre le altre librerie supportano eventuali analisi numeriche e visualizzazioni preliminari.


In [3]:
import pandas as p 
import numpy as np
import statistics as st
import matplotlib.pyplot as plt

## **Lettura di dati da file read.csv**

Il file Hotel_Reviews.csv viene caricato in un dataframe pandas per consentire l’analisi strutturata delle recensioni. Questo passaggio rappresenta il punto di partenza dell’intero processo di data cleaning, poiché permette di accedere alle osservazioni e alle variabili contenute nel dataset.

In [4]:
url= r"C:\Users\anton\Documents\Lavoro\Hotel_Reviews.csv"

hr=p.read_csv(url)

## **Fase EDA e DATA CLEANING**

Dopo il caricamento, il dataframe viene esplorato nelle sue caratteristiche principali per verificarne dimensioni, struttura e tipologie di dato. Questa fase preliminare consente di ottenere una visione d’insieme del dataset e di individuare eventuali criticità da affrontare nella pulizia. 

### Informazioni generali

Il controllo di shape e info(), permette di confermare il numero di osservazioni e variabili presenti nel dataset, oltre a evidenziare il tipo di dato associato a ciascuna colonna. Questa verifica è utile per identificare le variabili che richiedono conversioni o trasformazioni. 

In [5]:
hr.info()

<class 'pandas.DataFrame'>
RangeIndex: 515738 entries, 0 to 515737
Data columns (total 17 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   Hotel_Address                               515738 non-null  str    
 1   Additional_Number_of_Scoring                515738 non-null  int64  
 2   Review_Date                                 515738 non-null  str    
 3   Average_Score                               515738 non-null  float64
 4   Hotel_Name                                  515738 non-null  str    
 5   Reviewer_Nationality                        515738 non-null  str    
 6   Negative_Review                             515738 non-null  str    
 7   Review_Total_Negative_Word_Counts           515738 non-null  int64  
 8   Total_Number_of_Reviews                     515738 non-null  int64  
 9   Positive_Review                             515738 non-null  str    
 10  Review_

In [6]:
hr.shape

(515738, 17)

In particolare, il dataframe contiene 515.738 osservazioni e 17 variabili, a conferma della presenza di un archivio molto ampio di recensioni alberghiere.

### **Tipi di dato**

In [7]:
hr.dtypes

Hotel_Address                                     str
Additional_Number_of_Scoring                    int64
Review_Date                                       str
Average_Score                                 float64
Hotel_Name                                        str
Reviewer_Nationality                              str
Negative_Review                                   str
Review_Total_Negative_Word_Counts               int64
Total_Number_of_Reviews                         int64
Positive_Review                                   str
Review_Total_Positive_Word_Counts               int64
Total_Number_of_Reviews_Reviewer_Has_Given      int64
Reviewer_Score                                float64
Tags                                              str
days_since_review                                 str
lat                                           float64
lng                                           float64
dtype: object

L’analisi dei tipi di dato evidenzia la presenza di variabili testuali, numeriche intere e numeriche decimali. La colonna Review_Date è inizialmente memorizzata come stringa e dovrà essere convertita in datetime, mentre lat e lng risultano già in formato numerico decimale, utile per eventuali analisi geografiche.

### **Descrizione sul tipo di dato**

Con hr.describe(include='all') viene effettuata una descrizione generale del dataset, includendo sia le variabili numeriche sia quelle testuali. Per le colonne numeriche vengono restituiti indicatori come media, deviazione standard, minimo, massimo e quartili, mentre per le variabili di tipo stringa vengono mostrati elementi come numero di valori osservati, numero di valori distinti, valore più frequente e frequenza del valore più ricorrente.

Questa operazione consente di ottenere una visualizzazione sintetica dell’intero dataframe, utile per individuare distribuzioni, concentrazioni di valori e possibili anomalie prima di procedere con la fase di pulizia e trasformazione dei dati.

In [8]:
hr.describe(include='all')

,Hotel_Address,Additional_Number_of_Scoring,Review_Date,Average_Score,Hotel_Name,Reviewer_Nationality,Negative_Review,Review_Total_Negative_Word_Counts,Total_Number_of_Reviews,Positive_Review,Review_Total_Positive_Word_Counts,Total_Number_of_Reviews_Reviewer_Has_Given,Reviewer_Score,Tags,days_since_review,lat,lng
count,515738,515738.000000,515738,515738.000000,515738,515738,515738,515738.000000,515738.000000,515738,515738.000000,515738.000000,515738.000000,515738,515738,512470.000000,512470.000000
unique,1493,NaN,731,NaN,1492,227,330011,NaN,NaN,412601,NaN,NaN,NaN,55242,731,NaN,NaN
top,163 Marsh Wall Docklands Tower Hamlets London ...,NaN,8/2/2017,NaN,Britannia International Hotel Canary Wharf,United Kingdom,No Negative,NaN,NaN,No Positive,NaN,NaN,NaN,"[' Leisure trip ', ' Couple ', ' Double Room '...",1 days,NaN,NaN
freq,4789,NaN,2585,NaN,4789,245246,127890,NaN,NaN,35946,NaN,NaN,NaN,5101,2585,NaN,NaN
mean,NaN,498.081836,NaN,8.397487,NaN,NaN,NaN,18.539450,2743.743944,NaN,17.776458,7.166001,8.395077,NaN,NaN,49.442439,2.823803
std,NaN,500.538467,NaN,0.548048,NaN,NaN,NaN,29.690831,2317.464868,NaN,21.804185,11.040228,1.637856,NaN,NaN,3.466325,4.579425
min,NaN,1.000000,NaN,5.200000,NaN,NaN,NaN,0.000000,43.000000,NaN,0.000000,1.000000,2.500000,NaN,NaN,41.328376,-0.369758
25%,NaN,169.000000,NaN,8.100000,NaN,NaN,NaN,2.000000,1161.000000,NaN,5.000000,1.000000,7.500000,NaN,NaN,48.214662,-0.143372
50%,NaN,341.000000,NaN,8.400000,NaN,NaN,NaN,9.000000,2134.000000,NaN,11.000000,3.000000,8.800000,NaN,NaN,51.499981,0.010607
75%,NaN,660.000000,NaN,8.800000,NaN,NaN,NaN,23.000000,3613.000000,NaN,22.000000,8.000000,9.600000,NaN,NaN,51.516288,4.834443


### **Valori mancanti**

Viene effettuato il controllo dei valori mancanti, così da individuare eventuali colonne incomplete. 

In [9]:
hr.isnull().sum()

Hotel_Address                                    0
Additional_Number_of_Scoring                     0
Review_Date                                      0
Average_Score                                    0
Hotel_Name                                       0
Reviewer_Nationality                             0
Negative_Review                                  0
Review_Total_Negative_Word_Counts                0
Total_Number_of_Reviews                          0
Positive_Review                                  0
Review_Total_Positive_Word_Counts                0
Total_Number_of_Reviews_Reviewer_Has_Given       0
Reviewer_Score                                   0
Tags                                             0
days_since_review                                0
lat                                           3268
lng                                           3268
dtype: int64

L’analisi mostra che i missing values si concentrano nelle variabili geografiche lat e lng, mentre le altre colonne risultano sostanzialmente complete.

### **Ricerca dei duplicati**

Viene poi verificata la presenza di duplicati esatti nel dataset. Questo passaggio è importante per evitare che osservazioni replicate alterino le statistiche descrittive e i risultati delle analisi successive. Il controllo evidenzia la presenza di righe duplicate, che verranno rimosse nella fase di pulizia.

In [10]:
hr.duplicated().sum()

np.int64(526)

### **Distribuzione Reviewer** 

L’esplorazione iniziale include anche l’osservazione della distribuzione di Reviewer_Score e Average_Score, due variabili centrali per l’analisi. Entrambe mostrano una forte concentrazione su valori elevati, suggerendo una generale prevalenza di recensioni positive nel dataset.

In [11]:
print(hr['Reviewer_Score'].value_counts().sort_index())

Reviewer_Score
2.5       2199
2.9       1589
3.0         37
3.1          8
3.3       2772
3.5         79
3.8       4044
4.0         84
4.2       5095
4.4         12
4.5        122
4.6       6240
5.0       8422
5.4       9526
5.5        227
5.6         17
5.8      12034
6.0        253
6.3      14924
6.5        345
6.7      18793
6.9         29
7.0        411
7.1      24855
7.5      34878
7.9      33097
8.0        505
8.1         39
8.3      41128
8.5        508
8.8      46429
9.0        647
9.2      58583
9.4         63
9.5        724
9.6      71167
10.0    115853
Name: count, dtype: int64


La distribuzione di Reviewer_Score evidenzia che i punteggi assegnati dai recensori non sono distribuiti in modo uniforme lungo tutta la scala, ma risultano fortemente concentrati nelle fasce più alte. In particolare, i valori più frequenti si collocano tra 8 e 10, con un picco molto marcato sul punteggio 10.0, che rappresenta il valore più ricorrente nel dataset.

I punteggi più bassi, compresi tra circa 2.5 e 5.5, sono presenti ma con frequenze nettamente inferiori rispetto alle valutazioni alte. Questo suggerisce una generale tendenza dei recensori a esprimere giudizi positivi sulle strutture alberghiere analizzate. La distribuzione conferma che il dataset è caratterizzato da una prevalenza di recensioni favorevoli, mentre le valutazioni molto negative risultano meno comuni.

### **Distribuzione Average**

In [12]:
print(hr['Average_Score'].value_counts().sort_index())

Average_Score
5.2       65
6.4     1163
6.6      400
6.7      965
6.8     1335
6.9     1737
7.0     3899
7.1     6780
7.2      684
7.3     3997
7.4     6300
7.5    13720
7.6     6684
7.7    11829
7.8    19895
7.9    15194
8.0    22342
8.1    38122
8.2    34847
8.3    32880
8.4    41222
8.5    38066
8.6    36945
8.7    37798
8.8    30836
8.9    28520
9.0    21051
9.1    21379
9.2    12935
9.3    12659
9.4     9339
9.5     1207
9.6      915
9.8       28
Name: count, dtype: int64


La distribuzione di Average_Score mostra una forte concentrazione dei valori nella fascia medio-alta della scala, in particolare tra 8 e 8.7. I punteggi medi più bassi, inferiori a 7, risultano poco frequenti, mentre i valori estremamente elevati, vicini a 9.8, sono presenti ma in numero molto ridotto.

Rispetto a Reviewer_Score, la distribuzione di Average_Score appare più compatta e meno dispersa. Questo è coerente con la natura della variabile, che rappresenta un indicatore sintetico della reputazione media dell’hotel e non il giudizio del singolo recensore. Nel complesso, la distribuzione suggerisce che la maggior parte degli hotel del dataset presenta una valutazione media positiva, con una concentrazione marcata attorno ai valori centrali-alti.

In [13]:

distribuzione = hr[['Reviewer_Score','Average_Score']].describe()
print(distribuzione)

       Reviewer_Score  Average_Score
count   515738.000000  515738.000000
mean         8.395077       8.397487
std          1.637856       0.548048
min          2.500000       5.200000
25%          7.500000       8.100000
50%          8.800000       8.400000
75%          9.600000       8.800000
max         10.000000       9.800000


Le due variabili differiscono per dispersione. Reviewer_Score ha una deviazione standard più elevata (1.64) rispetto a Average_Score (0.55), indicando che i punteggi individuali sono molto più variabili rispetto alla media dell’hotel. Questo risultato è coerente con la natura delle due misure. Il punteggio del singolo recensore può oscillare maggiormente, mentre lo score medio dell’hotel è più stabile perché sintetizza molte recensioni.

### **Creo una copia di lavoro**

Per evitare modifiche dirette al dataset originale, viene creata una copia del dataframe. In questo modo, tutte le operazioni di pulizia e trasformazione vengono effettuate su una versione di lavoro separata, preservando i dati raw iniziali.

In [14]:
hr_clean = hr.copy()

### **Conversione in Datetime ed Estrazione Anno/Mese**

La colonna Review_Date viene convertita in formato datetime, così da rendere possibile l’estrazione di informazioni temporali in modo corretto. Questa operazione è necessaria per costruire variabili derivate utili alle analisi sull’andamento delle recensioni nel tempo.

A partire dalla data della recensione vengono estratte le variabili Review_Year e Review_Month. Queste nuove colonne consentono di effettuare analisi temporali più dettagliate, sia in SQL sia in pandas, come il conteggio delle recensioni per mese e per anno.  

In [15]:
hr_clean['Review_Date']  = p.to_datetime(hr_clean['Review_Date'])
hr_clean['Review_Year']  = hr_clean['Review_Date'].dt.year
hr_clean['Review_Month'] = hr_clean['Review_Date'].dt.month

hr_clean.dtypes 

Hotel_Address                                            str
Additional_Number_of_Scoring                           int64
Review_Date                                   datetime64[us]
Average_Score                                        float64
Hotel_Name                                               str
Reviewer_Nationality                                     str
Negative_Review                                          str
Review_Total_Negative_Word_Counts                      int64
Total_Number_of_Reviews                                int64
Positive_Review                                          str
Review_Total_Positive_Word_Counts                      int64
Total_Number_of_Reviews_Reviewer_Has_Given             int64
Reviewer_Score                                       float64
Tags                                                     str
days_since_review                                        str
lat                                                  float64
lng                     

### **Rimozione duplicati esatti**

Una volta identificati, i duplicati esatti vengono rimossi dal dataframe pulito. Questo garantisce che ciascuna riga rappresenti una recensione unica e migliora l’affidabilità delle elaborazioni successive.  

In [16]:
hr_clean.drop_duplicates(inplace=True)

In [17]:
hr_clean.dropna(inplace=True)  

hr_clean.isnull().sum()

Hotel_Address                                 0
Additional_Number_of_Scoring                  0
Review_Date                                   0
Average_Score                                 0
Hotel_Name                                    0
Reviewer_Nationality                          0
Negative_Review                               0
Review_Total_Negative_Word_Counts             0
Total_Number_of_Reviews                       0
Positive_Review                               0
Review_Total_Positive_Word_Counts             0
Total_Number_of_Reviews_Reviewer_Has_Given    0
Reviewer_Score                                0
Tags                                          0
days_since_review                             0
lat                                           0
lng                                           0
Review_Year                                   0
Review_Month                                  0
dtype: int64

### **Verifica che siano stati rimossi i duplicati**

In [18]:
hr_clean.duplicated().sum()

np.int64(0)

 Al termine di questo passaggio, la base dati risulta pronta per la creazione di nuove variabili e per il successivo caricamento nel database SQL.

### **Creazione delle variabili derivate**

###  Creazione della prima variabile derivata: tot recensioni per durata del soggiorno

Viene quindi creata la variabile stay_nights, ricavata dal campo Tags, per rappresentare il numero di notti di soggiorno dichiarato dal recensore. Questa nuova informazione arricchisce il dataset e può essere utile per analizzare il comportamento degli utenti in relazione ai punteggi assegnati.

In [19]:
hr_clean['stay_nights'] = hr_clean['Tags'].str.extract(r'Stayed\s+(\d+)\s+nights?')[0]

hr_clean['stay_nights'] = p.to_numeric(hr_clean['stay_nights'], errors='coerce').astype('Int64')

print(hr_clean['stay_nights'].value_counts().sort_index())

stay_nights
1     192757
2     132998
3      94865
4      47259
5      20619
6       9677
7       7324
8       2481
9       1283
10       881
11       416
12       315
13       242
14       246
15       117
16        57
17        42
18        33
19        29
20        19
21        22
22        10
23        10
24         6
25         7
26         6
27        10
28         6
29         5
30        10
31         1
Name: count, dtype: Int64


### Creazione della seconda variabile derivata: Lunghezza totale della recensione

Successivamente viene costruita la variabile total_review_length, ottenuta come somma della lunghezza della recensione positiva e della recensione negativa. Questo indicatore misura l’estensione complessiva del testo e sarà utile per le analisi dedicate alla relazione tra lunghezza della recensione e score.

In [20]:
hr_clean['total_review_length'] = (

    hr_clean['Positive_Review'].str.len() +

    hr_clean['Negative_Review'].str.len()

)

print(hr_clean['total_review_length'])   

0         1913
1          622
2          301
3         1221
4          774
          ... 
515733      69
515734      87
515735      88
515736     136
515737      66
Name: total_review_length, Length: 511944, dtype: int64


In [21]:
hr_clean.info()

<class 'pandas.DataFrame'>
Index: 511944 entries, 0 to 515737
Data columns (total 21 columns):
 #   Column                                      Non-Null Count   Dtype         
---  ------                                      --------------   -----         
 0   Hotel_Address                               511944 non-null  str           
 1   Additional_Number_of_Scoring                511944 non-null  int64         
 2   Review_Date                                 511944 non-null  datetime64[us]
 3   Average_Score                               511944 non-null  float64       
 4   Hotel_Name                                  511944 non-null  str           
 5   Reviewer_Nationality                        511944 non-null  str           
 6   Negative_Review                             511944 non-null  str           
 7   Review_Total_Negative_Word_Counts           511944 non-null  int64         
 8   Total_Number_of_Reviews                     511944 non-null  int64         
 9   Positive_

Una volta completate le operazioni di pulizia e trasformazione, il dataframe finale viene salvato come hotel_reviews_clean.csv. Questa versione pulita costituisce la base dati ufficiale da utilizzare nelle analisi successive con SQL, pandas e Matplotlib.

### **Salvataggio del dataset pulito in file csv**

In [22]:
hr_clean.to_csv('hotel_reviews_clean.csv', index=False)